File: src/algorithms/sorting/network/odd-even-merge-sort/sources/OddEvenMergeSort.cpp
// Odd-Even Merge Sort — Batcher's odd-even transposition sort (correct for all sizes)
#include <vector>
#include <algorithm>
#include <cmath>

std::vector<int> oddEvenMergeSort(std::vector<int> inputArray) {
    // @step:initialize
    std::vector<int> sortedArray = inputArray; // @step:initialize
    int arrayLength = sortedArray.size(); // @step:initialize

    if (arrayLength <= 1) {
        return sortedArray; // @step:complete
    }

    // Batcher's odd-even transposition sort:
    // Alternates between odd-phase and even-phase compare-swap passes
    // Requires ceil(n/2) * 2 rounds to sort n elements
    int totalRounds = (int)std::ceil(arrayLength / 2.0) * 2; // @step:merge

    for (int roundIndex = 0; roundIndex < totalRounds; roundIndex++) {
        // @step:compare
        bool isOddRound = roundIndex % 2 == 0; // @step:compare
        int startIndex = isOddRound ? 0 : 1; // @step:compare

        for (int leftIndex = startIndex; leftIndex + 1 < arrayLength; leftIndex += 2) {
            // @step:compare
            if (sortedArray[leftIndex] > sortedArray[leftIndex + 1]) {
                // @step:swap
                std::swap(sortedArray[leftIndex], sortedArray[leftIndex + 1]); // @step:swap
            }
        }
    }

    // @step:mark-sorted

    return sortedArray; // @step:complete
}

File: src/algorithms/sorting/network/odd-even-merge-sort/sources/odd-even-merge-sort.rs
// Odd-Even Merge Sort — Batcher's odd-even transposition sort (correct for all sizes)
fn odd_even_merge_sort(input_array: &[i64]) -> Vec<i64> {
    // @step:initialize
    let mut sorted_array = input_array.to_vec(); // @step:initialize
    let array_length = sorted_array.len(); // @step:initialize

    if array_length <= 1 {
        return sorted_array; // @step:complete
    }

    // Batcher's odd-even transposition sort:
    // Alternates between odd-phase and even-phase compare-swap passes
    // Requires ceil(n/2) * 2 rounds to sort n elements
    let total_rounds = ((array_length + 1) / 2) * 2; // @step:merge

    for round_index in 0..total_rounds {
        // @step:compare
        let is_odd_round = round_index % 2 == 0; // @step:compare
        let start_index = if is_odd_round { 0 } else { 1 }; // @step:compare

        let mut left_index = start_index;
        while left_index + 1 < array_length {
            // @step:compare
            if sorted_array[left_index] > sorted_array[left_index + 1] {
                // @step:swap
                sorted_array.swap(left_index, left_index + 1); // @step:swap
            }
            left_index += 2;
        }
    }

    // @step:mark-sorted

    sorted_array // @step:complete
}
